libraries/hotstuff/include/eosio/hotstuff/chain_pacemaker.hpp

```diff
@@ -15,6 +16,146 @@ namespace eosio::chain {
 
 namespace eosio::hotstuff {
 
+   class commitment_mgr_t {
+   public:
+      commitment_mgr_t(controller* chain) : _chain(chain) {
+      }
+
+      // called from net_plugin thread - must be synchronized
+      void store_commitment(const hs_commitment& commitment) {
+         fc::lock_guard g(_m);
+         _new_commitments.push_back(commitment);
+      }
+
+      // called from main thread
+      void get_new_commitments() {
+         fc::lock_guard g(_m);
+         for (auto& c : _new_commitments) {
+            const hs_proposal_message& p = c.b;
+            uint32_t block_num = block_header::num_from_id(p.block_id);
+            _commitments[block_num] = std::move(c);
+         }
+         _new_commitments.clear();
+      }
+            
+      // called from main thread
+      void push_required_commitment(const block_state_ptr& blk) {
+         uint32_t block_num = block_header::num_from_id(blk->id);
+         bool success = push_commitment(block_num);
+         assert(success);
+      }
+      
+      // called from main thread
+      void push_optional_commitment(const block_state_ptr& blk) {
+         uint32_t block_num = block_header::num_from_id(blk->id);
+         if (block_num - _last_pushed_commitment > 64)
```

Comment:
We should create a constant for this.

Reply:
Local to the class, or visible externally?

Reply:
Local to class is fine or even local to function. Just so it is obvious it is arbitrary.